src/Simplex/Messaging/Agent/RetryInterval.hs

```diff
@@ -63,6 +65,26 @@ withRetryIntervalCount ri action = callAction 0 0 $ initialInterval ri
           let elapsed' = elapsed + delay
           callAction (n + 1) elapsed' $ nextRetryDelay elapsed' delay ri
 
+withRetryForeground :: forall m a. MonadIO m => RetryInterval -> STM Bool -> STM Bool -> (Int64 -> m a -> m a) -> m a
+withRetryForeground ri isForeground isOnline action = callAction 0 $ initialInterval ri
+  where
+    callAction :: Int64 -> Int64 -> m a
+    callAction elapsed delay = action delay loop
+      where
+        loop = do
+          -- limit delay to max Int value (~36 minutes on for 32 bit architectures)
+          d <- registerDelay $ fromIntegral $ min delay (fromIntegral (maxBound :: Int))
+          (wasForground, wasOnline) <- atomically $ (,) <$> isForeground <*> isOnline
+          reset <- atomically $ do
+            foreground <- isForeground
+            online <- isOnline
+            let reset = (not wasForground && foreground) || (not wasOnline && online)
+            reset <$ unlessM ((reset ||) <$> readTVar d) retry
+          let (elapsed', delay')
+                | reset = (0, initialInterval ri)
+                | otherwise = (elapsed + delay, nextRetryDelay elapsed' delay ri)
+          callAction elapsed' delay'
+
```

Comment:
```suggestion
withRetryForeground :: forall m a. MonadIO m => RetryInterval -> STM Bool -> STM Bool -> (Int64 -> m a -> m a) -> m a
withRetryForeground ri isForeground isOnline action = callAction 0 $ initialInterval ri
  where
    callAction :: Int64 -> Int64 -> m a
    callAction elapsed delay = action delay loop
      where
        loop = do
          -- limit delay to max Int value (~36 minutes on for 32 bit architectures)
          d <- registerDelay $ fromIntegral $ min delay (fromIntegral (maxBound :: Int))
          (wasForeground, wasOnline) <- atomically $ (,) <$> isForeground <*> isOnline
          reset <- atomically $ do
            foreground <- isForeground
            online <- isOnline
            let reset = (not wasForeground && foreground) || (not wasOnline && online)
            reset <$ unlessM ((reset ||) <$> readTVar d) retry
          let (elapsed', delay')
                | reset = (0, initialInterval ri)
                | otherwise = (elapsed + delay, nextRetryDelay elapsed' delay ri)
          callAction elapsed' delay'

```